### PR TITLE
homework 3 done

### PR DESCRIPTION
**Pull request by PIDAMI**

### CODE REVIEW

File: src/test/java/com/epam/tc/hw3/page/objects/composite/components/BottomPartComponent.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+    @FindBy(css = "div.row.clerafix.benefits > div.col-sm-3 > "
+        + "div.benefit > span.benefit-txt")
+    private List<WebElement> textAfterImages;
+
+    public BottomPartComponent(WebDriver driver) {
+        super(driver);
+    }
+
+    public List<WebElement> getImages() {
+        return List.copyOf(images);
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/page/objects/composite/components/BottomPartComponent.java]: What is the purpose of List.copyOf?

File: src/test/java/com/epam/tc/hw3/page/objects/composite/components/LeftSectionComponent.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+import org.openqa.selenium.support.FindBy;
+
+public class LeftSectionComponent extends AbstractBaseComponent {
+
+    @FindBy(css = "ul.sidebar-menu.left > li > a > span")
+    List<WebElement> leftSectionElements;
+
+    public LeftSectionComponent(WebDriver driver) {
+
+        super(driver);
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/page/objects/composite/components/LeftSectionComponent.java:15]: Remove all extra lines all over the code!

File: src/test/java/com/epam/tc/hw3/page/objects/composite/components/ProfileComponent.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+        loginForm.sendKeys(user.getLogin());
+        loginButton.click();
+    }
+
+
+    public Optional<String> getUserNameIfDisplayed() {
+        return userName.isDisplayed()
+            ? Optional.of(userName.getText())
+            : Optional.empty();
+    }
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/page/objects/composite/components/ProfileComponent.java:43]: What is the purpose of using Optional here? Just use waiter or throw an exception.

File: src/test/java/com/epam/tc/hw3/page/objects/fluents/JdiIndexFluentPage.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    private final BottomPartComponent bottomPart;
+    private final LeftSectionComponent leftSection;
+    private final String indexPageUrl = "https://jdi-testing.github.io/jdi-light/index.html";
+
+    @FindBy(tagName = "iframe")
+    private List<WebElement> frames;
+
+    @FindBy(xpath = "//*[@class='uui-navigation nav navbar-nav m-l8']"
+        + "//*[contains(text(),'Service')]")
+    private WebElement serviceButton;
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/page/objects/fluents/JdiIndexFluentPage.java]: Isn't it enough to use just //*[contains(text(),'Service')] as locator?

File: src/test/java/com/epam/tc/hw3/ex2/SecondExerciseTest.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        Optional<String> userName = indexPage.profile().getUserNameIfDisplayed();
+        assertThat(userName.isPresent()).isTrue();
+        assertThat(userName.get()).isEqualTo(user.getDisplayedName());
+
+        // 5. Select checkboxes
+        DifferentElementsFluentPage differentElementsPage = indexPage.gotoDifferentElementPage();
+        List<String> elementsCheckboxesText = data.getElementsCheckboxesText();
+        List<Boolean> checkboxesChecked = differentElementsPage
+            .selectElementsCheckboxesAndCheckIfSelected(elementsCheckboxesText);
+        checkboxesChecked.forEach(checked -> assertThat(checked).isTrue());
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/ex2/SecondExerciseTest.java]: Very strange check. Don't use such way with list of Boolean. Check selected or not directly inside method.

File: src/test/java/com/epam/tc/hw3/ex2/SecondExerciseTest.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+        List<String> elementsLog = differentElementsPage
+            .getElementsCheckboxLogIfDisplayed(elementsCheckboxesText);
+        assertThat(elementsLog.size()).isEqualTo(elementsCheckboxesText.size());
+
+        // 9. Assert that
+        // for radio button there is a log row
+        // and value is corresponded to the status of radio button
+        Optional<String> radioLog = differentElementsPage
+            .getRadioCheckboxLogIfDisplayed(radioCheckboxesText);
+        assertThat(radioLog.isPresent()).isTrue();
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/ex2/SecondExerciseTest.java]: Use assertThat(radioLog).isPresent(); instead

File: src/test/java/com/epam/tc/hw3/page/objects/composite/components/BottomPartComponent.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.stream.Collectors;
+import org.openqa.selenium.WebDriver;
+import org.openqa.selenium.WebElement;
+import org.openqa.selenium.support.FindBy;
+
+public class BottomPartComponent extends AbstractBaseComponent {
+
+
+    @FindBy(css = "div.row.clerafix.benefits > div.col-sm-3 > "
+        + "div.benefit > div.benefit-icon > span.icons-benefit")
```
Comment (Delnote) [inline src/test/java/com/epam/tc/hw3/page/objects/composite/components/BottomPartComponent.java]: I know that locator can be way more simple. That one and the same are too weak and depend on lots of classes and DOM structure.

**Review by Delnote — COMMENTED** [review]
Pay attention to locators, simplify some options in methods.

File: src/test/java/com/epam/tc/hw3/page/objects/fluents/JdiIndexFluentPage.java (hunk 4 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw3/page/objects/fluents/JdiIndexFluentPage.java]: 2 different web elements have 'Service' as their text, so it's necessary to have such precise locators

File: src/test/java/com/epam/tc/hw3/page/objects/composite/components/BottomPartComponent.java (hunk 1 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw3/page/objects/composite/components/BottomPartComponent.java]: Not sure why i did it, changed to just `return images;`

File: src/test/java/com/epam/tc/hw3/page/objects/composite/components/LeftSectionComponent.java (hunk 2 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw3/page/objects/composite/components/LeftSectionComponent.java:15]: Done

File: src/test/java/com/epam/tc/hw3/page/objects/composite/components/ProfileComponent.java (hunk 3 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw3/page/objects/composite/components/ProfileComponent.java:43]: Changed to throwing an exception if name's not displayed

File: src/test/java/com/epam/tc/hw3/ex2/SecondExerciseTest.java (hunk 5 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw3/ex2/SecondExerciseTest.java]: Initially i didn't want to put test logic in page objects methods so it was done like this.
I agree, this does look strange, so I've moved asserts inside the methods

File: src/test/java/com/epam/tc/hw3/ex2/SecondExerciseTest.java (hunk 6 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw3/ex2/SecondExerciseTest.java]: Changed `Optional<String>` to `List<String>` and now assert that list's not empty

File: src/test/java/com/epam/tc/hw3/page/objects/composite/components/BottomPartComponent.java (hunk 7 above)
Comment (PIDAMI) [reply src/test/java/com/epam/tc/hw3/page/objects/composite/components/BottomPartComponent.java]: Changed to `span.icons-benefit`

**Comment by Delnote** [comment]
Ok, done. Can be merged.

**Pull request merged**